k2/csrc/rm_epsilon.cu

```diff
@@ -0,0 +1,210 @@
+      GetIncomingArcs(non_epsilon_fsa, GetDestStates(non_epsilon_fsa, true));
+
+  // epsilon_prec_renumbering will remap the arcs in epsilon_closure_mapped to
+  // the subset of arcs that we'll combine with the *preceding* arc (i.e.
+  // entering their src_state).
+  Renumbering epsilon_prec_renumbering(c, epsilon_closure_mapped.NumElements());
+  char *epsilon_prec_renumbering_keep_data =
+      epsilon_prec_renumbering.Keep().Data();
+
+  Array1<int32_t> epsilon_num_foll_arcs(
+      c, epsilon_closure_mapped.NumElements() + 1);
+  int32_t *epsilon_num_foll_arcs_data = epsilon_num_foll_arcs.Data();
+
+  // Lambda:
+  //   For each epsilon arc in epsilon_closure_mapped, we'll decide whether to
+  //   combine it with *following* non-epsilon arcs or *preceding* non-epsilon
+  //   arcs. We combine it with *following* non-epsilon arcs if it is leaving
+  //   from the start-state or if the num-non-epsilon-arcs leaving its dest
+  //   state is less than the num-non-epsilon-arcs entering its src state.
+  //
+  //   If we decided to combine it with following non-epsilon arcs then we set
+  //   epsilon_num_foll_arcs_data to the number of non-epsilon-arcs leaving
+  //   the dest-state, and set epsilon_prec_renumbering_keep_data to 0.
+  //   Else (combining with preceding arcs) we set epsilon_num_foll_arcs_data to
+  //   0 and set epsilon_prec_renumbering_keep_data to 1.
+
+  // `combined_foll` will be set to an FSA, with the same state numbering as
+  // `non_epsilon_fsa`, containing the arcs which arose by combining epsilon
+  // arcs with non-epsilon arcs following them.
+  FsaVec combined_foll;
+  Ragged<int32_t> combined_foll_arc_map;
+  {  // This block will set combined_foll and combined_foll_arc_map
+    ExclusiveSum(epsilon_num_foll_arcs, &epsilon_num_foll_arcs);
+    Array1<int32_t> &foll_row_splits = epsilon_num_foll_arcs;
+    int32_t num_arcs = foll_row_splits.Back();
+    Array1<int32_t> foll_row_ids(c, num_arcs);
+    RowSplitsToRowIds(foll_row_splits, &foll_row_ids);
+    // This shape just says, for each arc in epsilon_closure_mapped
+    // that is to be combined with following arcs, how many following
+    // arcs it is combined with (else 0).
+    RaggedShape foll_shape =
+        RaggedShape2(&foll_row_splits, &foll_row_ids, num_arcs);
+
+    // foll_non_eps_arc_idx will be set in the lambda to the arc-index within
+    // non_epsilon_fsa of the following arc which we're combining this epsilon
+    // arc with
+    Array1<int32_t> foll_non_eps_arc_idx(c, num_arcs);
+    Array1<Arc> arcs(c, num_arcs);
+    {
+      // lambda that sets foll_non_eps_arc_idx and arcs.
+    }
+
+    RaggedShape epsilon_closure_combined_with_fool =
+        ComposeRaggedShapes(epsilon_closure_mapped.shape, foll_shape);
+    RaggedShape foll_fsa_shape =
+        RemoveAxis(epsilon_closure_combined_with_fool, 2);
+
+    combined_foll = FsaVec(foll_fsa_shape, arcs);
+
+    Ragged<int32_t> epsilon_closure_mapped_arc_map_foll =
```

Comment:
@danpovey , as some of Ragged Ops require rvalue instead of lvalue (e.g. `AddSuffixToRagged`, `RemoveAxis`), I created some intermediate variables here. I just name them with some LONG BAD name for now to pass build ASAP, you may want to change them.

Reply:
Adding the vars is, I wasn't sure if that would compile, you can change as you see fit.